src/main/java/uk/gov/justice/digital/hmpps/keyworker/events/EventListener.java

```diff
@@ -1,53 +1,109 @@
-        final var event = getOffenderEvent(requestJson);
-        if (event != null) {
-            if ("EXTERNAL_MOVEMENT_RECORD-INSERTED".equals(event.getEventType())) {
+    public void eventListener(final String requestJson) throws JsonProcessingException {
+        final var message = getMessage(requestJson);
+        final var eventType = message.getMessageAttributes().getEventType().getValue();
+        log.info("Processing message of type {}", eventType);
+
+        final var event = getOffenderEvent(message.getMessage());
+        switch (eventType) {
+            case "EXTERNAL_MOVEMENT_RECORD-INSERTED":
                 reconciliationService.checkMovementAndDeallocate(event);
-            } else if ("BOOKING_NUMBER-CHANGED".equals(event.getEventType())) {
+                break;
+            case "BOOKING_NUMBER-CHANGED":
                 reconciliationService.checkForMergeAndDeallocate(event);
-            }
+                break;
+            case "DATA_COMPLIANCE_DELETE-OFFENDER":
+                Preconditions.checkState(StringUtils.isNotBlank(event.getOffenderIdDisplay()), String.format("Found null offender id for %s", requestJson));
+                keyworkerService.deleteKeyworkersForOffender(event.getOffenderIdDisplay());
+                break;
         }
     }
 
-    private OffenderEvent getOffenderEvent(final String requestJson) {
-        OffenderEvent event = null;
+    private Message getMessage(final String requestJson) throws JsonProcessingException {
         try {
-            final Map<String, String> message = objectMapper.readValue(requestJson, Map.class);
-            if (message != null && message.get("Message") != null) {
-                event = objectMapper.readValue(message.get("Message"), OffenderEvent.class);
-            }
-        } catch (IOException e) {
-            log.error("Failed to Parse Message {}", requestJson);
+            final var message = objectMapper.readValue(requestJson, new TypeReference<Message>() {
+            });
+            return Optional.ofNullable(message).orElseThrow(() -> new RuntimeException(String.format("Missing message from %s", requestJson)));
+        } catch (final JsonProcessingException e) {
+            log.error("Failed to parse Message {}", requestJson, e);
+            throw e;
+        }
+    }
+
+    private OffenderEvent getOffenderEvent(final String requestJson) throws JsonProcessingException {
+        try {
+            final var event = objectMapper.readValue(requestJson, new TypeReference<OffenderEvent>() {
+            });
+            return Optional.ofNullable(event).orElseThrow(() -> new RuntimeException(String.format("Missing offender event from %s", requestJson)));
+        } catch (final JsonProcessingException e) {
+            log.error("Failed to parse event {}", requestJson, e);
+            throw e;
+        }
+    }
+
+    @Getter
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    private static class Message {
```

Comment:
this is far too much code even with lombok.  The kotlin versions are single line each...

Reply:
Indeed.  How do you apply the Json annotations in Kotlin on one line?

Reply:
Java 14 adds a preview of records which have similar syntax to kotlin data classes: https://openjdk.java.net/jeps/359

Reply:
Operation Kill-Kotlin begins.